llvm/lib/IR/RepoTicket.cpp

```diff
@@ -78,186 +87,236 @@ const Constant *getAliasee(const GlobalAlias *GA) {
+    LLVM_DEBUG(dbgs() << "Returning pre-computed hash for \"" << GO->getName()
+                      << "\"\n");
+    return std::make_tuple(GOVisitedIndex, GODigest);
+  }
 
+  MD5 GOHash = MD5();
   bool Inserted;
   typename GOStateMap::const_iterator StateIt;
-  std::tie(StateIt, Inserted) = Visited.try_emplace(GO, Visited.size());
+  std::tie(StateIt, Inserted) = Visited.try_emplace(GO, GOVisitedIndex);
   if (!Inserted) {
-    // If GO is visited, use the letter 'R' as the marker and use its state as
-    // the value.
-    GOHash.update('R');
-    GOHash.update(StateIt->second);
-    return true;
+    assert(GOVisitedIndex > StateIt->second);
+    // If GO is visited, mark as a Backref and use its state as the value.
+    GOHash.update(static_cast<char>(Tags::Backref));
+    GOHash.update(GOVisitedIndex - StateIt->second - 1U);
+    LLVM_DEBUG(dbgs() << "Hashing back reference to #" << StateIt->second
+                      << "\n");
+    DigestType Digest;
+    GOHash.final(Digest);
+    Changed = true;
+    return std::make_tuple(StateIt->second, Digest);
   }
 
-  GOHash.update('T');
-  auto State = AccumulateGODigest(GO, GOHash, GOIMap);
-  bool Changed = State.Changed;
+  GOHash.update(static_cast<char>(Tags::GO));
+  auto State = (this->*AccumulateGODigest)(GO, GOHash);
   // The hashes of the GO's 'contributions' and 'dependencies' are separately
   // added to the GO's hash since their respective arrows point are in different
   // direction. In addition, 'contributions' cannot form loops since 1) only
   // global variables have 'contributions' and 2) all 'contributions' are
   // functions. Therefore, the 'contributions' can be simplified not to record
   // 'visited' table, which avoids having two 'visited' maps. Firstly, add the
   // GO's contributions to its hash.
-  Changed = updateDigestUseContributions(GOHash, State.Contributions, GOIMap) ||
-            Changed;
+  updateDigestUseContributions(GOHash, State.Contributions);
 
   // Add the GO's dependencies to its hash.
-  Changed = updateDigestUseDependencies(GOHash, State.Dependencies, Visited,
-                                        GOIMap, AccumulateGODigest) ||
-            Changed;
-
-  return Changed;
+  return updateDigestUseDependencies<Function>(
+      GO, GOHash, GOVisitedIndex, State.Dependencies, AccumulateGODigest);
 }
 
-// Calculate the GOs' initial digest, dependencies, contributions and GONumber.
-std::pair<GOInfoMap, GONumber> calculateGONumAndGOIMap(Module &M) {
-  GOInfoMap GOIMap;
-  GONumber GONum;
-
-  for (auto &GV : M.globals()) {
-    if (GV.isDeclaration())
-      continue;
-    calculateGOInfo(&GV, GOIMap);
-    ++GONum.VarNum;
-  }
-  for (auto &Fn : M.functions()) {
-    if (Fn.isDeclaration())
-      continue;
-    calculateGOInfo(&Fn, GOIMap);
-    ++GONum.FuncNum;
-  }
-  return std::make_pair(std::move(GOIMap), std::move(GONum));
+template <typename Function>
+DigestType HashCache::calculateDigest(const GlobalObject *GO,
```

Comment:
I feel like there is a significant functional difference and usage between:
```
calculateDigest(const GlobalObject *GO, Function AccumulateGODigest)
calculateDigest(const GlobalObject *GO)
```

That is not well communicated by being overloaded functions.

The same goes for:

```
HashCache()
HashCache(Module &M)
```

Reply:
>```
> calculateDigest(const GlobalObject *GO)
> ```
>
> That is not well communicated by being overloaded functions. Should they have different names?

Renamed this to `calculateDigestUseRepoDefinition(const GlobalObject *GO)`.


> 
> Should the latter be a method (e.g. digestModule()) to be called after construction?
> 

Changed, thanks.